<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from gazebo_65_description.urdf.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="rm_65_description">
  <link name="base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.000433277303987328 -3.54664423471128E-05 0.0599427668933796"/>
      <mass value="0.841070778135659"/>
      <!-- <inertia
        ixx="0.0017261110801622"
        ixy="2.52746264980217E-06"
        ixz="-3.67690303614961E-05"
        iyy="0.00170987405835604"
        iyz="1.67996364994424E-06"
        izz="0.000904023422915791" />
    </inertial> -->
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <!-- <mesh filename="file://$(find rm_description)/meshes/rm_65_arm/base_link.STL" /> -->
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/base_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/base_link.STL"/>
      </geometry>
    </collision>
  </link>
  <link name="Link1">
    <inertial>
      <origin rpy="0 0 0" xyz="1.2226305431569E-08 0.0211079974844683 -0.0251854220842269"/>
      <mass value="0.593563443690403"/>
      <!-- <inertia
        ixx="0.00126614120341847"
        ixy="-1.294980943835E-08"
        ixz="-9.80120923066996E-09"
        iyy="0.00118168178300364"
        iyz="-0.00021121727444415"
        izz="0.00056135241627747" />
    </inertial> -->
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link1.STL"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link1.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint1" type="revolute">
    <origin rpy="0 0 0" xyz="0 0 0.2405"/>
    <parent link="base_link"/>
    <child link="Link1"/>
    <axis xyz="0 0 1"/>
    <limit effort="60" lower="-3.1" upper="3.1" velocity="3.14"/>
  </joint>
  <link name="Link2">
    <inertial>
      <origin rpy="0 0 0" xyz="0.152256463426163 4.75383656106654E-07 -0.00620260383607792"/>
      <mass value="0.864175046869043"/>
      <!-- <inertia
        ixx="0.00089150298478414"
        ixy="-2.23268489334765E-08"
        ixz="0.00156246461035015"
        iyy="0.00733754343083901"
        iyz="6.28110889329165E-09"
        izz="0.00697869103915473" />
    </inertial> -->
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link2.STL"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link2.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint2" type="revolute">
    <origin rpy="1.5708 -1.5708 0" xyz="0 0 0"/>
    <parent link="Link1"/>
    <child link="Link2"/>
    <axis xyz="0 0 1"/>
    <limit effort="60" lower="-2.268" upper="2.268" velocity="3.14"/>
  </joint>
  <link name="Link3">
    <inertial>
      <origin rpy="0 0 0" xyz="5.05312670989961E-06 -0.0595925663694732 0.010569069212336"/>
      <mass value="0.289633681624654"/>
      <!-- <inertia
        ixx="0.00063737100450158"
        ixy="-7.05261293649751E-08"
        ixz="-3.86643272239426E-08"
        iyy="0.00015648388095025"
        iyz="-0.00014461035994916"
        izz="0.000614178164773085" />
    </inertial> -->
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link3.STL"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link3.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint3" type="revolute">
    <origin rpy="0 0 1.5708" xyz="0.256 0 0"/>
    <parent link="Link2"/>
    <child link="Link3"/>
    <axis xyz="0 0 1"/>
    <limit effort="30" lower="-2.355" upper="2.355" velocity="3.92"/>
  </joint>
  <link name="Link4">
    <inertial>
      <origin rpy="0 0 0" xyz="1.15516617405898E-06 -0.0180424468598241 -0.0215394748352687"/>
      <mass value="0.239419768320061"/>
      <!-- <inertia
        ixx="0.000285938919722783"
        ixy="3.07101359163101E-09"
        ixz="-2.21994118981953E-09"
        iyy="0.000262727540304212"
        iyz="4.4236583260078E-05"
        izz="0.000119888082791859" /> 
    </inertial>-->
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link4.STL"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link4.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint4" type="revolute">
    <origin rpy="1.5708 0 0" xyz="0 -0.21 0"/>
    <parent link="Link3"/>
    <child link="Link4"/>
    <axis xyz="0 0 1"/>
    <limit effort="10" lower="-3.1" upper="3.1" velocity="3.92"/>
  </joint>
  <link name="Link5">
    <inertial>
      <origin rpy="0 0 0" xyz="3.19794786262152E-06 -0.0593808368101458 0.00736804250989326"/>
      <mass value="0.218799761431678"/>
      <!-- <inertia
        ixx="0.000350540363914072"
        ixy="-3.41781619975602E-08"
        ixz="-1.77056457224373E-08"
        iyy="0.000104927867487581"
        iyz="-7.82431228461971E-05"
        izz="0.000334482418423629" />
    </inertial> -->
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link5.STL"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link5.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint5" type="revolute">
    <origin rpy="-1.5708 0 0" xyz="0 0 0"/>
    <parent link="Link4"/>
    <child link="Link5"/>
    <axis xyz="0 0 1"/>
    <limit effort="10" lower="-2.233" upper="2.233" velocity="3.92"/>
  </joint>
  <link name="Link6">
    <inertial>
      <origin rpy="0 0 0" xyz="0.000714234511756999 -0.000396718732824521 -0.0126723660946126"/>
      <mass value="0.0649018034311231"/>
      <!-- <inertia
        ixx="2.02766547502765E-05"
        ixy="-1.32505200276849E-06"
        ixz="-2.5845091522508E-08"
        iyy="1.87986725225022E-05"
        iyz="3.39471452125439E-09"
        izz="3.17885459163081E-05" />
    </inertial> -->
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link6.STL"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/rm_65_arm/link6.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="joint6" type="revolute">
    <origin rpy="1.5708 0 0" xyz="0 -0.144 0"/>
    <parent link="Link5"/>
    <child link="Link6"/>
    <axis xyz="0 0 1"/>
    <limit effort="10" lower="-6.28" upper="6.28" velocity="3.92"/>
  </joint>
  <!-- Used for fixing robot to Gazebo 'base_link' 将机械手的基座固定在世界坐标上-->
  <link name="world"/>
  <joint name="world-base-fix" type="fixed">
    <parent link="world"/>
    <child link="base_link"/>
  </joint>
  <!--定义一个工具坐标系-->
  <!-- <link name="gripper_tip_link"/> -->
  <!-- <joint name="gripper_tip_joint" type="fixed"> -->
  <!-- <origin xyz="0 0 0.1523" rpy="0 0 0"/> -->
  <!-- <origin xyz="0 0 0" rpy="0 0 0"/> -->
  <!-- <parent link="Link6"/> -->
  <!-- <child link="gripper_tip_link"/> -->
  <!-- </joint> -->
  <!-- gripper -->
  <link name="gripper_base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="-2.6792E-06 1.4343E-05 -0.040955"/>
      <mass value="0.075349"/>
      <inertia ixx="6.9695E-05" ixy="-2.8955E-12" ixz="1.2653E-08" iyy="6.0426E-05" iyz="2.2585E-08" izz="1.9195E-05"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/gripper/gripper_base_link.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.79216 0.81961 0.93333 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/gripper/gripper_base_link.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="arm_joint_gripper" type="fixed">
    <!-- <origin xyz="0 0 0.008" rpy="0 0 0" /> -->
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="Link6"/>
    <child link="gripper_base_link"/>
  </joint>
  <link name="gripper_Link1">
    <inertial>
      <origin rpy="0 0 0" xyz="0.0001 0.0085832 0.013664"/>
      <mass value="0.0024857"/>
      <inertia ixx="5.8907E-07" ixy="1.3343E-19" ixz="4.2984E-19" iyy="5.2237E-07" iyz="-2.5787E-07" izz="2.7386E-07"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/gripper/gripper_Link1.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.79216 0.81961 0.93333 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/gripper/gripper_Link1.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="gripper_joint1" type="revolute">
    <origin rpy="0.8663 0 0" xyz="0 -0.04 0.09"/>
    <parent link="gripper_base_link"/>
    <child link="gripper_Link1"/>
    <axis xyz="1 0 0"/>
    <limit effort="100" lower="-0.8" upper="0.6" velocity="1"/>
  </joint>
  <link name="gripper_Link2">
    <inertial>
      <origin rpy="0 0 0" xyz="9.9997E-05 0.011846 0.011367"/>
      <mass value="0.0044927"/>
      <inertia ixx="4.099E-07" ixy="-7.8736E-14" ixz="-7.8248E-14" iyy="4.2946E-07" iyz="-3.8654E-08" izz="1.6162E-07"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/gripper/gripper_Link2.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.79216 0.81961 0.93333 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/gripper/gripper_Link2.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="gripper_joint2" type="revolute">
    <origin rpy="-0.8663 0 0" xyz="0 0.022341 0.035565"/>
    <parent link="gripper_Link1"/>
    <child link="gripper_Link2"/>
    <axis xyz="-1 0 0"/>
    <limit effort="100" lower="-0.8" upper="0.6" velocity="1"/>
  </joint>
  <link name="gripper_Link3">
    <inertial>
      <origin rpy="0 0 0" xyz="0.0001 -0.0085832 0.013664"/>
      <mass value="0.0024857"/>
      <inertia ixx="5.8907E-07" ixy="-3.3324E-19" ixz="3.1193E-19" iyy="5.2237E-07" iyz="2.5787E-07" izz="2.7386E-07"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/gripper/gripper_Link3.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.79216 0.81961 0.93333 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/gripper/gripper_Link3.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="gripper_joint3" type="revolute">
    <origin rpy="-0.8663 0 0" xyz="0 0.04 0.09"/>
    <parent link="gripper_base_link"/>
    <child link="gripper_Link3"/>
    <axis xyz="-1 0 0"/>
    <limit effort="100" lower="-0.8" upper="0.6" velocity="1"/>
  </joint>
  <link name="gripper_Link4">
    <inertial>
      <origin rpy="0 0 0" xyz="9.9999E-05 -0.011846 0.011367"/>
      <mass value="0.0044926"/>
      <inertia ixx="4.099E-07" ixy="-1.4237E-13" ixz="-2.483E-13" iyy="4.2946E-07" iyz="3.8655E-08" izz="1.6162E-07"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/gripper/gripper_Link4.STL"/>
      </geometry>
      <material name="">
        <color rgba="0.79216 0.81961 0.93333 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/txszju/Code/addx/ws_rm_robot/install/rm_description/share/rm_description/meshes/gripper/gripper_Link4.STL"/>
      </geometry>
    </collision>
  </link>
  <joint name="gripper_joint4" type="revolute">
    <origin rpy="0.8663 0 0" xyz="0 -0.022341 0.035565"/>
    <parent link="gripper_Link3"/>
    <child link="gripper_Link4"/>
    <axis xyz="1 0 0"/>
    <limit effort="100" lower="-0.8" upper="0.6" velocity="1"/>
  </joint>
  <!-- 对一些link进行gazebo的属性设置 -->
  <gazebo reference="base_link">
    <self_collide>false</self_collide>
    <gravity>false</gravity>
  </gazebo>
  <gazebo reference="Link1">
    <material>Gazebo/Purple</material>
    <self_collide>false</self_collide>
    <gravity>false</gravity>
  </gazebo>
  <gazebo reference="Link2">
    <material>Gazebo/Red</material>
    <gravity>false</gravity>
  </gazebo>
  <gazebo reference="Link3">
    <material>Gazebo/Blue</material>
    <gravity>false</gravity>
  </gazebo>
  <gazebo reference="Link4">
    <material>Gazebo/Green</material>
    <gravity>false</gravity>
  </gazebo>
  <gazebo reference="Link5">
    <material>Gazebo/Yellow</material>
    <gravity>false</gravity>
  </gazebo>
  <gazebo reference="Link6">
    <material>Gazebo/Orange</material>
    <gravity>false</gravity>
  </gazebo>
  <gazebo>
    <is_static>true</is_static>
    <self_collide>true</self_collide>
  </gazebo>
  <!-- 声明ros2_control -->
  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="joint1">
      <command_interface name="position"/>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">0</param>
      </state_interface>
      <state_interface name="velocity">
        <param name="initial_value">0</param>
      </state_interface>
    </joint>
    <joint name="joint2">
      <command_interface name="position"/>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">0</param>
      </state_interface>
      <state_interface name="velocity">
        <param name="initial_value">0</param>
      </state_interface>
    </joint>
    <joint name="joint3">
      <command_interface name="position"/>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">0</param>
      </state_interface>
      <state_interface name="velocity">
        <param name="initial_value">0</param>
      </state_interface>
    </joint>
    <joint name="joint4">
      <command_interface name="position"/>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">0</param>
      </state_interface>
      <state_interface name="velocity">
        <param name="initial_value">0</param>
      </state_interface>
    </joint>
    <joint name="joint5">
      <command_interface name="position"/>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">0</param>
      </state_interface>
      <state_interface name="velocity">
        <param name="initial_value">0</param>
      </state_interface>
    </joint>
    <joint name="joint6">
      <command_interface name="position"/>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">0</param>
      </state_interface>
      <state_interface name="velocity">
        <param name="initial_value">0</param>
      </state_interface>
    </joint>
    <joint name="gripper_joint1">
      <command_interface name="position"/>
      <!-- <command_interface name="velocity"/> -->
      <state_interface name="position">
        <param name="initial_value">0</param>
      </state_interface>
      <state_interface name="velocity">
        <param name="initial_value">0</param>
      </state_interface>
    </joint>
    <joint name="gripper_joint2">
      <command_interface name="position"/>
      <!-- <command_interface name="velocity"/> -->
      <state_interface name="position">
        <param name="initial_value">0</param>
      </state_interface>
      <state_interface name="velocity">
        <param name="initial_value">0</param>
      </state_interface>
    </joint>
    <joint name="gripper_joint3">
      <command_interface name="position"/>
      <!-- <command_interface name="velocity"/> -->
      <state_interface name="position">
        <param name="initial_value">0</param>
      </state_interface>
      <state_interface name="velocity">
        <param name="initial_value">0</param>
      </state_interface>
    </joint>
    <joint name="gripper_joint4">
      <command_interface name="position"/>
      <!-- <command_interface name="velocity"/> -->
      <state_interface name="position">
        <param name="initial_value">0</param>
      </state_interface>
      <state_interface name="velocity">
        <param name="initial_value">0</param>
      </state_interface>
    </joint>
  </ros2_control>
  <!-- 加载ros2_control插件 -->
  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <parameters>/home/txszju/Code/addx/ws_rm_robot/install/rm_65_config/share/rm_65_config/config/ros2_controllers.yaml</parameters>
      <robot_param>robot_description</robot_param>
      <robot_param_node>robot_state_publisher</robot_param_node>
    </plugin>
  </gazebo>
</robot>
